python/tvm/relay/transform/quantization/quantized_operators/dense.py

```diff
@@ -0,0 +1,183 @@
+from typing import *
+
+import numpy as np
+import tvm
+from tvm import relay
+from tvm.relay.dataflow_pattern import (
+    DFPatternCallback,
+    _DFPatternCallback,
+    is_constant,
+    is_op,
+    wildcard,
+)
+from tvm.relay.op import nn, tensor
+from tvm.relay.transform.quantization import calibration_rewrite_utils
+
+QParams = NamedTuple(
+    "QParams", [("scale_factor", tvm.relay.Expr), ("zero_point", tvm.relay.Expr), ("dtype", str)]
+)
+
+
+class AffineQuantizationVarCreator:
+    """Class which manages references to our qparams and can insert state."""
+
+    def __init__(self):
+        self.ref_count = 0
+        self.qparams = []
+
+    def get_qparams(self, name_hint: str, dtype: str = "int8") -> QParams:
+        scale = relay.Var(f"{name_hint}.scale")
+        zero_point = relay.Var(f"{name_hint}.zero_point")
+        qparam = QParams(scale, zero_point, dtype)
+        self.qparams.append(qparam)
+        self.ref_count += 1
+        return qparam
+
+
+def generate_generic_quantized_dense(
+    data: tvm.relay.Expr,
+    weight: tvm.relay.Expr,
+    data_qparams: QParams,
+    weight_qparams: QParams,
+    internal_accumulation_dtype: str = "float32",
+    simulated_accumulation_dtype: str = "int32",
+    out_units: Optional[int] = None,
+    in_units: Optional[int] = None,
+) -> Tuple[tvm.relay.Expr, QParams]:
+    """TODO"""
+
+    # TODO: figure out whether we need this or we can always have the
+    # callee pass it in
+    if in_units is None:
+        units_in = weight.checked_type.shape[-1]
+    if out_units is None:
+        out_units = weight.checked_type.shape[-2]
+
+    data_scale, data_zero_point, data_dtype = data_qparams
+    weight_scale, weight_zero_point, weight_dtype = weight_qparams
+
+    # Assume this casting is a no-op in the case we are casting back to itself
+    weight_zero_point = relay.cast(weight_zero_point, internal_accumulation_dtype)
+    data_zero_point = relay.cast(data_zero_point, internal_accumulation_dtype)
+
+    data = relay.cast(data, internal_accumulation_dtype)
+    weight = relay.cast(weight, internal_accumulation_dtype)
+
+    first_term = nn.dense(data, weight, units=out_units, out_dtype=internal_accumulation_dtype)
+
+    # The fields for the reduction operations to make things clear
+    axis = [1]
+    keep_dims = True
+    exclude = False
+
+    second_term = tensor._make.sum(data, axis, keep_dims, exclude) * weight_zero_point
```

Comment:
Here, you're using an internal constructor to create the sum. I think it's better practice to use `relay.op.sum` instead. Also that way, you don't need to pass axis in as an array, you can pass it in as an integer. 

https://github.com/apache/tvm/blob/main/python/tvm/relay/op/reduce.py#L88

Reply:
done good tip